File: src/sdk/cpp/ofa_agent/src/p2p.cpp
/**
 * @file p2p.cpp
 * @brief P2P Communication Implementation
 * Sprint 30: P2P Enhancement
 */

#include "ofa/p2p.hpp"
#include <chrono>
#include <thread>
#include <random>
#include <sstream>

// Platform-specific socket includes
#ifdef _WIN32
    #include <winsock2.h>
    #include <ws2tcpip.h>
    #pragma comment(lib, "ws2_32.lib")
    #define CLOSE_SOCKET closesocket
    #define SOCKET_TYPE SOCKET
    #define INVALID_SOCKET_VAL INVALID_SOCKET
#else
    #include <sys/socket.h>
    #include <netinet/in.h>
    #include <arpa/inet.h>
    #include <unistd.h>
    #define CLOSE_SOCKET close
    #define SOCKET_TYPE int
    #define INVALID_SOCKET_VAL -1
#endif

namespace ofa {

// Initialize Winsock on Windows
#ifdef _WIN32
static bool winsockInitialized = false;
static void initWinsock() {
    if (!winsockInitialized) {
        WSADATA wsaData;
        WSAStartup(MAKEWORD(2, 2), &wsaData);
        winsockInitialized = true;
    }
}
#endif

// Helper functions
static std::string generateId() {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(0, 15);

    std::stringstream ss;
    ss << std::hex;
    for (int i = 0; i < 16; i++) ss << dis(gen);
    return ss.str();
}

static int64_t nowMs() {
    return std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::system_clock::now().time_since_epoch()
    ).count();
}

// ==================== P2PClient ====================

P2PClient::P2PClient(const std::string& agentId,
                     uint16_t listenPort,
                     uint16_t discoveryPort)
    : agentId_(agentId)
    , listenPort_(listenPort)
    , discoveryPort_(discoveryPort) {
#ifdef _WIN32
    initWinsock();
#endif
}

P2PClient::~P2PClient() {
    stop();
}

std::string P2PClient::generateId() {
    return ::ofa::generateId();
}

std::future<bool> P2PClient::start() {
    return std::async(std::launch::async, [this]() {
        if (running_) return true;

        // Create TCP socket
        tcpSocket_ = socket(AF_INET, SOCK_STREAM, 0);
        if (tcpSocket_ == INVALID_SOCKET_VAL) {
            return false;
        }

        // Bind to port
        struct sockaddr_in addr;
        addr.sin_family = AF_INET;
        addr.sin_addr.s_addr = INADDR_ANY;
        addr.sin_port = htons(listenPort_ == 0 ? 7891 : listenPort_);

        if (bind(tcpSocket_, (struct sockaddr*)&addr, sizeof(addr)) < 0) {
            CLOSE_SOCKET(tcpSocket_);
            return false;
        }

        listen(tcpSocket_, 5);

        // Get actual port
        if (listenPort_ == 0) {
            socklen_t len = sizeof(addr);
            getsockname(tcpSocket_, (struct sockaddr*)&addr, &len);
            listenPort_ = ntohs(addr.sin_port);
        }

        // Create UDP socket for discovery
        udpSocket_ = socket(AF_INET, SOCK_DGRAM, 0);
        if (udpSocket_ == INVALID_SOCKET_VAL) {
            CLOSE_SOCKET(tcpSocket_);
            return false;
        }

        // Enable broadcast
        int broadcast = 1;
        setsockopt(udpSocket_, SOL_SOCKET, SO_BROADCAST,
                   (const char*)&broadcast, sizeof(broadcast));

        running_ = true;
        tcpThread_ = std::thread(&P2PClient::tcpListen, this);

        return true;
    });
}

void P2PClient::stop() {
    if (!running_) return;
    running_ = false;
    discovering_ = false;

    if (tcpSocket_ != INVALID_SOCKET_VAL) {
        CLOSE_SOCKET(tcpSocket_);
        tcpSocket_ = INVALID_SOCKET_VAL;
    }

    if (udpSocket_ != INVALID_SOCKET_VAL) {
        CLOSE_SOCKET(udpSocket_);
        udpSocket_ = INVALID_SOCKET_VAL;
    }

    if (tcpThread_.joinable()) {
        tcpThread_.join();
    }
    if (udpThread_.joinable()) {
        udpThread_.join();
    }
    if (timeoutThread_.joinable()) {
        timeoutThread_.join();
    }
}

void P2PClient::startDiscovery(const std::vector<std::string>& skills) {
    if (discovering_) return;
    discovering_ = true;
    udpThread_ = std::thread(&P2PClient::udpDiscovery, this);
    timeoutThread_ = std::thread(&P2PClient::checkPeerTimeout, this);
}

void P2PClient::stopDiscovery() {
    discovering_ = false;
}

std::vector<PeerInfo> P2PClient::getPeers() const {
    std::lock_guard<std::mutex> lock(mutex_);
    std::vector<PeerInfo> result;
    for (const auto& pair : peers_) {
        if (pair.second.isOnline) {
            result.push_back(pair.second);
        }
    }
    return result;
}

PeerInfo P2PClient::getPeer(const std::string& peerId) const {
    std::lock_guard<std::mutex> lock(mutex_);
    auto it = peers_.find(peerId);
    if (it == peers_.end()) {
        throw OfAException("Peer not found: " + peerId);
    }
    return it->second;
}

std::future<bool> P2PClient::send(const std::string& peerId, const P2PMessage& msg) {
    return std::async(std::launch::async, [this, peerId, msg]() {
        std::lock_guard<std::mutex> lock(mutex_);

        auto it = peers_.find(peerId);
        if (it == peers_.end() || !it->second.isOnline) {
            return false;
        }

        // Create socket and connect
        SOCKET_TYPE sock = socket(AF_INET, SOCK_STREAM, 0);
        if (sock == INVALID_SOCKET_VAL) {
            return false;
        }

        struct sockaddr_in addr;
        addr.sin_family = AF_INET;
        addr.sin_port = htons(it->second.port);

#ifdef _WIN32
        InetPtonA(AF_INET, it->second.address.c_str(), &addr.sin_addr);
#else
        inet_pton(AF_INET, it->second.address.c_str(), &addr.sin_addr);
#endif

        if (connect(sock, (struct sockaddr*)&addr, sizeof(addr)) < 0) {
            CLOSE_SOCKET(sock);
            return false;
        }

        // Send message
        std::string data = msg.toJson().dump();
        send(sock, data.c_str(), (int)data.size(), 0);

        CLOSE_SOCKET(sock);
        return true;
    });
}

std::future<void> P2PClient::broadcast(const P2PMessage& msg) {
    return std::async(std::launch::async, [this, msg]() {
        auto peers = getPeers();
        for (const auto& peer : peers) {
            send(peer.id, msg).get();
        }
    });
}

std::future<json> P2PClient::sendTaskRequest(const std::string& peerId,
                                              const std::string& skillId,
                                              const std::string& operation,
                                              const json& input,
                                              int timeoutMs) {
    return std::async(std::launch::async, [this, peerId, skillId, operation, input, timeoutMs]() {
        P2PMessage msg;
        msg.id = generateId();
        msg.type = P2PMessageType::TaskRequest;
        msg.from = agentId_;
        msg.to = peerId;
        msg.data = {
            {"skillId", skillId},
            {"operation", operation},
            {"input", input}
        };
        msg.timestamp = nowMs();

        // In production: send and wait for response
        // For now: simulate
        send(peerId, msg).get();

        // Would wait for TaskResponse
        return json{{"status", "sent"}};
    });
}

void P2PClient::onMessage(P2PMessageType type, P2PMessageHandler handler) {
    std::lock_guard<std::mutex> lock(mutex_);
    handlers_[type].push_back(handler);
}

void P2PClient::onPeerOnline(std::function<void(const PeerInfo&)> callback) {
    onPeerOnline_ = callback;
}

void P2PClient::onPeerOffline(std::function<void(const PeerInfo&)> callback) {
    onPeerOffline_ = callback;
}

void P2PClient::tcpListen() {
    while (running_) {
        struct sockaddr_in clientAddr;
        socklen_t clientLen = sizeof(clientAddr);

        // Set non-blocking with timeout
        fd_set readfds;
        FD_ZERO(&readfds);
        FD_SET(tcpSocket_, &readfds);

        struct timeval tv;
        tv.tv_sec = 1;
        tv.tv_usec = 0;

        int activity = select((int)tcpSocket_ + 1, &readfds, nullptr, nullptr, &tv);

        if (activity <= 0 || !running_) continue;

        SOCKET_TYPE clientSock = accept(tcpSocket_, (struct sockaddr*)&clientAddr, &clientLen);
        if (clientSock == INVALID_SOCKET_VAL) continue;

        // Handle connection in thread
        std::thread(&P2PClient::handleConnection, this, clientSock).detach();
    }
}

void P2PClient::handleConnection(SOCKET_TYPE clientSock) {
    char buffer[4096];
    int bytesRead = recv(clientSock, buffer, sizeof(buffer) - 1, 0);

    if (bytesRead > 0) {
        buffer[bytesRead] = '\0';

        try {
            json j = json::parse(std::string(buffer, bytesRead));
            P2PMessage msg = P2PMessage::fromJson(j);
            processMessage(msg);
        } catch (...) {
            // Invalid message
        }
    }

    CLOSE_SOCKET(clientSock);
}

void P2PClient::processMessage(const P2PMessage& msg) {
    // Handle discovery
    if (msg.type == P2PMessageType::Discovery) {
        PeerInfo peer = PeerInfo::fromJson(msg.data);
        peer.lastSeen = nowMs();
        peer.isOnline = true;

        std::lock_guard<std::mutex> lock(mutex_);
        bool isNew = peers_.find(peer.id) == peers_.end();
        peers_[peer.id] = peer;

        if (isNew && onPeerOnline_) {
            onPeerOnline_(peer);
        }

        // Send ack
        P2PMessage ack;
        ack.id = generateId();
        ack.type = P2PMessageType::DiscoveryAck;
        ack.from = agentId_;
        ack.data = {{}};  // Self info would go here
        send(peer.id, ack);
    }

    // Handle task request
    if (msg.type == P2PMessageType::TaskRequest) {
        std::lock_guard<std::mutex> lock(mutex_);
        auto it = handlers_.find(P2PMessageType::TaskRequest);
        if (it != handlers_.end()) {
            for (auto& handler : it->second) {
                handler(msg);
            }
        }
    }

    // Call registered handlers
    std::lock_guard<std::mutex> lock(mutex_);
    auto it = handlers_.find(msg.type);
    if (it != handlers_.end()) {
        for (auto& handler : it->second) {
            handler(msg);
        }
    }
}

void P2PClient::checkPeerTimeout() {
    while (running_) {
        std::this_thread::sleep_for(std::chrono::seconds(5));

        auto now = nowMs();
        std::lock_guard<std::mutex> lock(mutex_);

        for (auto& pair : peers_) {
            if (now - pair.second.lastSeen > 30000) {  // 30s timeout
                if (pair.second.isOnline) {
                    pair.second.isOnline = false;
                    if (onPeerOffline_) {
                        onPeerOffline_(pair.second);
                    }
                }
            }
        }
    }
}

void P2PClient::udpDiscovery() {
    while (discovering_ && running_) {
        sendDiscoveryBroadcast();
        std::this_thread::sleep_for(std::chrono::seconds(5));
    }
}

void P2PClient::sendDiscoveryBroadcast() {
    P2PMessage msg;
    msg.id = generateId();
    msg.type = P2PMessageType::Discovery;
    msg.from = agentId_;
    msg.timestamp = nowMs();

    // Self info
    PeerInfo selfInfo;
    selfInfo.id = agentId_;
    selfInfo.address = "0.0.0.0";  // Would get actual IP
    selfInfo.port = listenPort_;
    msg.data = selfInfo.toJson();

    std::string data = msg.toJson().dump();

    struct sockaddr_in addr;
    addr.sin_family = AF_INET;
    addr.sin_port = htons(discoveryPort_);
#ifdef _WIN32
    InetPtonA(AF_INET, "255.255.255.255", &addr.sin_addr);
#else
    inet_pton(AF_INET, "255.255.255.255", &addr.sin_addr);
#endif

    sendto(udpSocket_, data.c_str(), (int)data.size(), 0,
           (struct sockaddr*)&addr, sizeof(addr));
}

// ==================== P2PDiscovery ====================

P2PDiscovery::P2PDiscovery(uint16_t port)
    : port_(port) {
#ifdef _WIN32
    initWinsock();
#endif
}

P2PDiscovery::~P2PDiscovery() {
    stop();
}

bool P2PDiscovery::start() {
    socket_ = socket(AF_INET, SOCK_DGRAM, 0);
    if (socket_ == INVALID_SOCKET_VAL) {
        return false;
    }

    int broadcast = 1;
    setsockopt(socket_, SOL_SOCKET, SO_BROADCAST,
               (const char*)&broadcast, sizeof(broadcast));

    struct sockaddr_in addr;
    addr.sin_family = AF_INET;
    addr.sin_addr.s_addr = INADDR_ANY;
    addr.sin_port = htons(port_);

    if (bind(socket_, (struct sockaddr*)&addr, sizeof(addr)) < 0) {
        CLOSE_SOCKET(socket_);
        return false;
    }

    running_ = true;
    thread_ = std::thread(&P2PDiscovery::listenLoop, this);
    return true;
}

void P2PDiscovery::stop() {
    running_ = false;
    if (socket_ != INVALID_SOCKET_VAL) {
        CLOSE_SOCKET(socket_);
    }
    if (thread_.joinable()) {
        thread_.join();
    }
}

void P2PDiscovery::broadcast(const PeerInfo& selfInfo) {
    P2PMessage msg;
    msg.id = generateId();
    msg.type = P2PMessageType::Discovery;
    msg.from = selfInfo.id;
    msg.data = selfInfo.toJson();
    msg.timestamp = nowMs();

    std::string data = msg.toJson().dump();

    struct sockaddr_in addr;
    addr.sin_family = AF_INET;
    addr.sin_port = htons(port_);
#ifdef _WIN32
    InetPtonA(AF_INET, "255.255.255.255", &addr.sin_addr);
#else
    inet_pton(AF_INET, "255.255.255.255", &addr.sin_addr);
#endif

    sendto(socket_, data.c_str(), (int)data.size(), 0,
           (struct sockaddr*)&addr, sizeof(addr));
}

void P2PDiscovery::onDiscovery(std::function<void(const PeerInfo&)> callback) {
    onDiscovery_ = callback;
}

void P2PDiscovery::listenLoop() {
    char buffer[4096];
    struct sockaddr_in clientAddr;
    socklen_t clientLen = sizeof(clientAddr);

    while (running_) {
        fd_set readfds;
        FD_ZERO(&readfds);
        FD_SET(socket_, &readfds);

        struct timeval tv;
        tv.tv_sec = 1;
        tv.tv_usec = 0;

        int activity = select(socket_ + 1, &readfds, nullptr, nullptr, &tv);

        if (activity <= 0 || !running_) continue;

        int bytesRead = recvfrom(socket_, buffer, sizeof(buffer) - 1, 0,
                                 (struct sockaddr*)&clientAddr, &clientLen);

        if (bytesRead > 0) {
            buffer[bytesRead] = '\0';

            try {
                json j = json::parse(std::string(buffer, bytesRead));
                P2PMessage msg = P2PMessage::fromJson(j);

                if (msg.type == P2PMessageType::Discovery) {
                    PeerInfo peer = PeerInfo::fromJson(msg.data);

                    // Get sender address
#ifdef _WIN32
                    char ipStr[INET_ADDRSTRLEN];
                    InetNtopA(AF_INET, &clientAddr.sin_addr, ipStr, INET_ADDRSTRLEN);
#else
                    char ipStr[INET_ADDRSTRLEN];
                    inet_ntop(AF_INET, &clientAddr.sin_addr, ipStr, INET_ADDRSTRLEN);
#endif

                    peer.address = ipStr;
                    peer.lastSeen = nowMs();
                    peer.isOnline = true;

                    if (onDiscovery_) {
                        onDiscovery_(peer);
                    }
                }
            } catch (...) {
                // Invalid message
            }
        }
    }
}

} // namespace ofa

File: src/sdk/go/p2p.go
package ofa

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net"
	"sync"
	"time"
)

// PeerInfo represents a peer node
type PeerInfo struct {
	ID        string            `json:"id"`
	Name      string            `json:"name"`
	Address   string            `json:"address"`
	Port      int               `json:"port"`
	Skills    []string          `json:"skills"`
	Metadata  map[string]string `json:"metadata"`
	LastSeen  int64             `json:"lastSeen"`
	IsOnline  bool              `json:"isOnline"`
}

// P2PMessageType defines P2P message type
type P2PMessageType int

const (
	P2PMsgDiscovery P2PMessageType = iota
	P2PMsgDiscoveryAck
	P2PMsgTaskRequest
	P2PMsgTaskResponse
	P2PMsgDataSync
	P2PMsgHeartbeat
	P2PMsgCustom
)

// P2PMessage represents a P2P message
type P2PMessage struct {
	ID        string          `json:"id"`
	Type      P2PMessageType  `json:"type"`
	From      string          `json:"from"`
	To        string          `json:"to"`
	Data      json.RawMessage `json:"data"`
	Timestamp int64           `json:"timestamp"`
}

// P2PMessageHandler handles P2P messages
type P2PMessageHandler func(*P2PMessage)

// P2PClient manages P2P communication
type P2PClient struct {
	agentID       string
	listenPort    int
	discoveryPort int

	peers    map[string]*PeerInfo
	handlers map[P2PMessageType][]P2PMessageHandler
	onPeerOnline  func(*PeerInfo)
	onPeerOffline func(*PeerInfo)

	mu       sync.Mutex
	running  bool
	done     chan struct{}

	tcpListener net.Listener
	udpConn     *net.UDPConn
}

// NewP2PClient creates a new P2P client
func NewP2PClient(agentID string, listenPort, discoveryPort int) *P2PClient {
	return &P2PClient{
		agentID:       agentID,
		listenPort:    listenPort,
		discoveryPort: discoveryPort,
		peers:         make(map[string]*PeerInfo),
		handlers:      make(map[P2PMessageType][]P2PMessageHandler),
		done:          make(chan struct{}),
	}
}

// Start starts the P2P client
func (p *P2PClient) Start() error {
	p.mu.Lock()
	if p.running {
		p.mu.Unlock()
		return nil
	}
	p.running = true
	p.mu.Unlock()

	// Start TCP listener
	if p.listenPort == 0 {
		p.listenPort = 7891
	}

	addr := fmt.Sprintf(":%d", p.listenPort)
	listener, err := net.Listen("tcp", addr)
	if err != nil {
		return fmt.Errorf("failed to start TCP listener: %w", err)
	}
	p.tcpListener = listener

	// Start UDP for discovery
	udpAddr, err := net.ResolveUDPAddr("udp", fmt.Sprintf(":%d", p.discoveryPort))
	if err != nil {
		udpAddr, err = net.ResolveUDPAddr("udp", ":0")
		if err != nil {
			return fmt.Errorf("failed to resolve UDP address: %w", err)
		}
	}

	udpConn, err := net.ListenUDP("udp", udpAddr)
	if err != nil {
		return fmt.Errorf("failed to start UDP listener: %w", err)
	}
	p.udpConn = udpConn

	// Start background loops
	go p.tcpAcceptLoop()
	go p.udpReceiveLoop()
	go p.peerTimeoutLoop()

	log.Printf("P2P client started on port %d", p.listenPort)
	return nil
}

// Stop stops the P2P client
func (p *P2PClient) Stop() {
	p.mu.Lock()
	p.running = false
	p.mu.Unlock()

	close(p.done)

	if p.tcpListener != nil {
		p.tcpListener.Close()
	}
	if p.udpConn != nil {
		p.udpConn.Close()
	}
}

// StartDiscovery starts peer discovery
func (p *P2PClient) StartDiscovery() {
	go p.discoveryLoop()
}

// GetPeers returns discovered peers
func (p *P2PClient) GetPeers() []*PeerInfo {
	p.mu.Lock()
	defer p.mu.Unlock()

	var peers []*PeerInfo
	for _, peer := range p.peers {
		if peer.IsOnline {
			peers = append(peers, peer)
		}
	}
	return peers
}

// GetPeer returns a specific peer
func (p *P2PClient) GetPeer(peerID string) (*PeerInfo, error) {
	p.mu.Lock()
	defer p.mu.Unlock()

	peer, ok := p.peers[peerID]
	if !ok {
		return nil, fmt.Errorf("peer not found: %s", peerID)
	}
	return peer, nil
}

// Send sends a message to a peer
func (p *P2PClient) Send(peerID string, msg *P2PMessage) error {
	p.mu.Lock()
	peer, ok := p.peers[peerID]
	if !ok || !peer.IsOnline {
		p.mu.Unlock()
		return fmt.Errorf("peer not online: %s", peerID)
	}
	p.mu.Unlock()

	// Connect to peer
	addr := fmt.Sprintf("%s:%d", peer.Address, peer.Port)
	conn, err := net.DialTimeout("tcp", addr, 5*time.Second)
	if err != nil {
		return fmt.Errorf("failed to connect to peer: %w", err)
	}
	defer conn.Close()

	// Send message
	data, err := json.Marshal(msg)
	if err != nil {
		return fmt.Errorf("failed to marshal message: %w", err)
	}

	_, err = conn.Write(data)
	return err
}

// Broadcast broadcasts to all peers
func (p *P2PClient) Broadcast(msg *P2PMessage) error {
	peers := p.GetPeers()
	for _, peer := range peers {
		if err := p.Send(peer.ID, msg); err != nil {
			log.Printf("Failed to send to peer %s: %v", peer.ID, err)
		}
	}
	return nil
}

// SendTaskRequest sends a task request to a peer
func (p *P2PClient) SendTaskRequest(ctx context.Context, peerID, skillID, operation string, input json.RawMessage, timeout time.Duration) (json.RawMessage, error) {
	msg := &P2PMessage{
		ID:        generateID(),
		Type:      P2PMsgTaskRequest,
		From:      p.agentID,
		To:        peerID,
		Timestamp: time.Now().UnixMilli(),
	}

	taskData := map[string]interface{}{
		"skillId":   skillID,
		"operation": operation,
		"input":     input,
	}
	msg.Data, _ = json.Marshal(taskData)

	if err := p.Send(peerID, msg); err != nil {
		return nil, err
	}

	// In production: wait for response
	// For now, return placeholder
	return json.RawMessage(`{"status":"sent"}`), nil
}

// OnMessage registers a message handler
func (p *P2PClient) OnMessage(msgType P2PMessageType, handler P2PMessageHandler) {
	p.mu.Lock()
	p.handlers[msgType] = append(p.handlers[msgType], handler)
	p.mu.Unlock()
}

// OnPeerOnline sets peer online callback
func (p *P2PClient) OnPeerOnline(callback func(*PeerInfo)) {
	p.mu.Lock()
	p.onPeerOnline = callback
	p.mu.Unlock()
}

// OnPeerOffline sets peer offline callback
func (p *P2PClient) OnPeerOffline(callback func(*PeerInfo)) {
	p.mu.Lock()
	p.onPeerOffline = callback
	p.mu.Unlock()
}

// LocalPort returns the local port
func (p *P2PClient) LocalPort() int {
	return p.listenPort
}

func (p *P2PClient) tcpAcceptLoop() {
	for {
		select {
		case <-p.done:
			return
		default:
			conn, err := p.tcpListener.Accept()
			if err != nil {
				if p.running {
					log.Printf("TCP accept error: %v", err)
				}
				continue
			}

			go p.handleTCPConnection(conn)
		}
	}
}

func (p *P2PClient) handleTCPConnection(conn net.Conn) {
	defer conn.Close()

	conn.SetReadDeadline(time.Now().Add(10 * time.Second))

	buf := make([]byte, 4096)
	n, err := conn.Read(buf)
	if err != nil {
		return
	}

	var msg P2PMessage
	if err := json.Unmarshal(buf[:n], &msg); err != nil {
		log.Printf("Failed to unmarshal message: %v", err)
		return
	}

	p.processMessage(&msg)
}

func (p *P2PClient) udpReceiveLoop() {
	buf := make([]byte, 4096)

	for {
		select {
		case <-p.done:
			return
		default:
			n, remoteAddr, err := p.udpConn.ReadFromUDP(buf)
			if err != nil {
				if p.running {
					log.Printf("UDP read error: %v", err)
				}
				continue
			}

			var msg P2PMessage
			if err := json.Unmarshal(buf[:n], &msg); err != nil {
				continue
			}

			// Handle discovery
			if msg.Type == P2PMsgDiscovery {
				var peer PeerInfo
				if err := json.Unmarshal(msg.Data, &peer); err != nil {
					continue
				}

				peer.Address = remoteAddr.IP.String()
				peer.LastSeen = time.Now().UnixMilli()
				peer.IsOnline = true

				p.mu.Lock()
				isNew := p.peers[peer.ID] == nil
				p.peers[peer.ID] = &peer

				if isNew && p.onPeerOnline != nil {
					go p.onPeerOnline(&peer)
				}
				p.mu.Unlock()

				// Send ack
				ack := &P2PMessage{
					ID:        generateID(),
					Type:      P2PMsgDiscoveryAck,
					From:      p.agentID,
					Timestamp: time.Now().UnixMilli(),
				}
				p.sendUDPAck(remoteAddr, ack)
			}
		}
	}
}

func (p *P2PClient) sendUDPAck(addr *net.UDPAddr, msg *P2PMessage) {
	data, _ := json.Marshal(msg)
	p.udpConn.WriteToUDP(data, addr)
}

func (p *P2PClient) discoveryLoop() {
	ticker := time.NewTicker(5 * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-p.done:
			return
		case <-ticker.C:
			p.sendDiscoveryBroadcast()
		}
	}
}

func (p *P2PClient) sendDiscoveryBroadcast() {
	selfInfo := &PeerInfo{
		ID:      p.agentID,
		Address: "0.0.0.0",
		Port:    p.listenPort,
	}

	msg := &P2PMessage{
		ID:        generateID(),
		Type:      P2PMsgDiscovery,
		From:      p.agentID,
		Timestamp: time.Now().UnixMilli(),
	}
	msg.Data, _ = json.Marshal(selfInfo)

	data, _ := json.Marshal(msg)

	// Broadcast to discovery port
	broadcastAddr := &net.UDPAddr{
		IP:   net.IPv4(255, 255, 255, 255),
		Port: p.discoveryPort,
	}

	if p.udpConn != nil {
		p.udpConn.WriteToUDP(data, broadcastAddr)
	}
}

func (p *P2PClient) peerTimeoutLoop() {
	ticker := time.NewTicker(5 * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-p.done:
			return
		case <-ticker.C:
			p.mu.Lock()
			now := time.Now().UnixMilli()
			for id, peer := range p.peers {
				if now - peer.LastSeen > 30000 { // 30s timeout
					peer.IsOnline = false
					if p.onPeerOffline != nil {
						go p.onPeerOffline(peer)
					}
					delete(p.peers, id)
				}
			}
			p.mu.Unlock()
		}
	}
}

func (p *P2PClient) processMessage(msg *P2PMessage) {
	p.mu.Lock()
	handlers := p.handlers[msg.Type]
	p.mu.Unlock()

	for _, handler := range handlers {
		go handler(msg)
	}
}